sycl/doc/SYCLPluginInterface.md

```diff
@@ -41,12 +40,45 @@ once before any actual offload is attempted.
 
 ### Plugin discovery
 
-Plugins are physically dynamic libraries stored somewhere in the system where
-the SYCL runtime runs. TBD - design and describe the process in details.
+Plugins are physically dynamic libraries or shared objects.
+The process to discover plugins will follow the following guidelines.
+
+The SYCL Runtime will search for plugins at env LD_LIBRARY_PATH location on
+Linux and env LIB on Windows, with names in the format of libpiXXX.so and
+query them. An extension to the search mechanism is to use a configuration file
+which lists all the available plugins and their locations. 
+This file can be stored along with SYCL Runtime.
+Plugins should expose the information of supported PI API version as a constant
+string, so it can be read without loading the library.
+The user can select/disable a specifc plugin with an environment variable.
+The Plugin Interface queries the plugins on the supported PI version and check
+for compatibility.
+The Plugin Interface then queries each plugin for all Interface Functions
+corresponding function pointers and populate a list of these pointers for
+each plugin. As an extension, the plugin can provide a populated function
+pointer table/struct for each device attached to the plugin.
+A trace mechanism should be provided to log the discovery/connection/device
+enumeration process. Eg: Display all the plugins being discovered, their
+information and supported PI version. List attached devices and their properties.
+
+ TBD - design and describe the process in detail.
 
 #### Plugin binary interface
-TBD - list and describe all the symbols plugin must export in order to be picked
-up by the SYCL runtime for offload.
+Currently the plugins should export all the symbols that are present in the file
+pi.h. In the future, this document will list the minimum set of Interface APIs
+to be supported by Plugins. This will also require adding functionality to SYCL
+Runtime to work with such limited functionality plugins.
+
+TBD - list and describe all the symbols that a plugin must export in order to
+be picked up by the SYCL runtime for offload.
+
+#### Connection to a Plugin
+Plugin selection can be forced via environment variable: SYCL_PI_USE.
+Default is OpenCL Plugin. The connection is established when PI is forwarding
```

Comment:
My initial understanding was that all plugins that listed in config are loaded. Are we going to support only one plugin at the same time?

Reply:
We are going to connect (bind) SYCL RT to single plugin/back-end.

Reply:
So only one plugin can be active at the same time? No interoperability across PI plugins is expected at the SYCL level? Note that this is fairly doable in SYCL

Reply:
> We are going to connect (bind) SYCL RT to single plugin/back-end.

@smaslov-intel, please clarify. User must be able to offload simultaneously to multiple devices and PI interface is the mechanism behind this.

Reply:
Sorry for misleading comments here. You are right, PI will be redirecting the SYCL RT requests to the established connections with possibly multiple PI plugins (covering possibly multiple devices). The de-multiplexing will be based on the actual device/context being used and requests routed to the PI plugin associated with the device/plugin.

Reply:
Thanks, sounds good.

Reply:
There are multiple alternatives how we could organize the routing. Suppose we have a table of interfaces (let's call it dispatch table) provided by each PI plugin, so we need a way to know which dispatch table to look at for every PI call. Few ways of doing this would be:

1) Have each PI handle to carry information about it's PI plugin dispatch table.
Pros:
+ PI handles are self-explanatory, and so PI calls can be routed outside of any SYCL context 
Cons:
- memory overhead
- extra translation from PI handles to OpenCL handles, which otherwise could be made equal
- the PI dispatch table is not generally known/used by the specific PI interfaces creating these PI handles

2) Keep the pointer to the PI plugin dispatch table in the SYCL objects (possibly just in a SYCL context) from where it can always be fetched at the time of a PI call.
Pros:
+ Probably less memory overhead at the price of possibly higher run-time overhead, where later is probably going to be negligible.
+ More natural to not keep data that you don't understand or use. SYCL RT continues to have the PI handles of opaque types.
Cons:
- You need to know the SYCL context (or other upper-level context if PI is ever used outside of SYCL) to route a PI call.

Thoughts, preferences?

Reply:
> --- extra translation from PI handles to OpenCL handles, which otherwise could be made equal
Could you elaborate more on this cons?

Reply:
> Could you elaborate more on this cons?

If we add extra stuff into PI handles, then there is more work to extract native OpenCL handle from a PI  handle. Note that today most of PI handles returned by PI OpenCL plugin are exactly the OpenCL native raw handles.

We are strongly in favor of approach 2, which is to keep/cache the PI plugin descriptors in the SYCL RT, not in the PI handles.

Reply:
Option 2 seems a good approach. The SYCL context is already used to hold many "state-related" information, and most SYCL API entries, including interop, require a SYCL context as a parameter.

Reply:
One thing that must be clarified is: A SYCL context selects the PI Plugin *on construction* and that never changes for said SYCL context.
Can't have a SYCL context switching plugins or mix-and-matching plugins.